Fix an issue when using nested scopedMDC

nkgeorgiev opened the pull request:
ScopedMDC is a RAII class that removes the value associated with its key from the MDC map when it goes out of scope.
But if we have nested ScopedMDC objects on the same key, the inner one overwrites the outer. 
When the inner object is destroyed, it deletes the value from the MDC map instead of restoring the one from the outer scope.
This leads to a loss of information inside the logs.

The fix is to store the previous value inside the ScopedMDC object and restore it on destruction.

toly-kournik commented:
What's the example of a nested MDC?
What if there are more than one nested MDC?

f-squirrel commented:
> What's the example of a nested MDC?
> What if there are more than one nested MDC?

From my understanding, every scoped MDC in a chain restores its previous value.
Is it correct @nkgeorgiev ?
However, I wonder why we have multiple MDC with the same value:)

nkgeorgiev commented:
> What's the example of a nested MDC?

One of the patterns is to create a scopedMDC at the start of onMessage methods.

Here for example:
https://github.com/vmware/concord-bft/blob/master/bftengine/src/bftengine/ReplicaImp.cpp#L2148
onMessage<ViewChangeMsg> sets the seqNum value to the view number (which is an issue in itself but it's harder to change the format of the logs now)
After that it calls onMessage<ReplicaAsksToLeaveViewMsg> which again sets the seqNum value. I changed that call to non-scoped MDC_PUT in a previous change as a temp fix. I have forgotten to revert it in this PR. I will update it in a minute :)

> What if there are more than one nested MDC?
It will become like a stack of values with each nested MDC pushing a value and popping when it is destroyed. The logger always gets the top value. Since MDC is thread local there should be no issues.

nkgeorgiev commented:
> > What's the example of a nested MDC?
> > What if there are more than one nested MDC?
> 
> From my understanding, every scoped MDC in a chain restores its previous value.
> Is it correct @nkgeorgiev ?
> However, I wonder why we have multiple MDC with the same key:)

I gave an example of mutiple MDC with the same key in my previous comments.
There are also many redundant MDC that were trying to workaround this problem.
I will start removing them in a different PR.

toly-kournik commented:
> > What's the example of a nested MDC?
> 
> One of the patterns is to create a scopedMDC at the start of onMessage methods.
> 
> Here for example:
> https://github.com/vmware/concord-bft/blob/master/bftengine/src/bftengine/ReplicaImp.cpp#L2148
> onMessage sets the seqNum value to the view number (which is an issue in itself but it's harder to change the format of the logs now)
> After that it calls onMessage which again sets the seqNum value. I changed that call to non-scoped MDC_PUT in a previous change as a temp fix. I have forgotten to revert it in this PR. I will update it in a minute :)
> 
> > What if there are more than one nested MDC?
> 
> It will become like a stack of values with each nested MDC pushing a value and popping when it is destroyed. The logger always gets the top value. Since MDC is thread local there should be no issues.



> > What's the example of a nested MDC?
> 
> One of the patterns is to create a scopedMDC at the start of onMessage methods.
> 
> Here for example:
> https://github.com/vmware/concord-bft/blob/master/bftengine/src/bftengine/ReplicaImp.cpp#L2148
> onMessage sets the seqNum value to the view number (which is an issue in itself but it's harder to change the format of the logs now)
> After that it calls onMessage which again sets the seqNum value. I changed that call to non-scoped MDC_PUT in a previous change as a temp fix. I have forgotten to revert it in this PR. I will update it in a minute :)
> 
> > What if there are more than one nested MDC?
> 
> It will become like a stack of values with each nested MDC pushing a value and popping when it is destroyed. The logger always gets the top value. Since MDC is thread local there should be no issues.

Using view number as a sequence number is a misuse!

nkgeorgiev commented:
> > > What's the example of a nested MDC?
> > 
> > 
> > One of the patterns is to create a scopedMDC at the start of onMessage methods.
> > Here for example:
> > https://github.com/vmware/concord-bft/blob/master/bftengine/src/bftengine/ReplicaImp.cpp#L2148
> > onMessage sets the seqNum value to the view number (which is an issue in itself but it's harder to change the format of the logs now)
> > After that it calls onMessage which again sets the seqNum value. I changed that call to non-scoped MDC_PUT in a previous change as a temp fix. I have forgotten to revert it in this PR. I will update it in a minute :)
> > > What if there are more than one nested MDC?
> > 
> > 
> > It will become like a stack of values with each nested MDC pushing a value and popping when it is destroyed. The logger always gets the top value. Since MDC is thread local there should be no issues.
> 
> > > What's the example of a nested MDC?
> > 
> > 
> > One of the patterns is to create a scopedMDC at the start of onMessage methods.
> > Here for example:
> > https://github.com/vmware/concord-bft/blob/master/bftengine/src/bftengine/ReplicaImp.cpp#L2148
> > onMessage sets the seqNum value to the view number (which is an issue in itself but it's harder to change the format of the logs now)
> > After that it calls onMessage which again sets the seqNum value. I changed that call to non-scoped MDC_PUT in a previous change as a temp fix. I have forgotten to revert it in this PR. I will update it in a minute :)
> > > What if there are more than one nested MDC?
> > 
> > 
> > It will become like a stack of values with each nested MDC pushing a value and popping when it is destroyed. The logger always gets the top value. Since MDC is thread local there should be no issues.
> 
> Using view number as a sequence number is a misuse!

I agree, but it is a different issue for another PR
